### ISSUE TITLE
Renaming scalars

### ISSUE THREAD

**Issue by richieshilton** [apollographql/apollo-ios#2798]
### Question

We're just wondering if it's possible to rename scalars on fetching the schema or during code generation (using  the Apollo iOS cli). I've had a look through the documentation and can't see any options available for this in the codegen config. 

I understand that we can setup custom scalar handling, but renaming them would be really useful if possible.

Thanks!

**Comment by brunnoferreira** [comment]
This is a good point, I'd like to second that.

We are currently facing an issue that could be easily overcome by renaming custom scalars – we build a client for an API which we don't control (and feel might not have been well planned) so there's many custom types that overlap and some of them with naming that clashes with Foundation types.

While namespacing _helps_, it adds a lot of noise to our codebase (let's say, to differentiate between `Foundation.Date` and `ApolloGenerated.Date`) for something that wasn't our choice to begin with.

**Comment by AnthonyMDev** [comment]
This is a future addition we do want to make. It's just not super high priority right now compared to some of the things we are working on.

As a side note, if you are using `Foundation.Date` as the underlying type for your custom scalar and you set the `typealias ApolloGenerated.Date = Foundation.Date`, you can just reference it in your code as `Date` and it shouldn't be a conflict. This is only a naming conflict if you have a separate concrete type for your custom scalar.

**Comment by calvincestari** [comment]
Closing due to there being no action we can take in Apollo iOS for this issue. If you believe you have a bug related to this issue please comment in this issue again.

**Comment by dafurman** [comment]
Our backend has used "Void" to represent Null values in our schema which wasn't a problem until Apollo 1.0. Note: We also have an identical issue for "URL".

```json
{
        "kind": "SCALAR",
        "name": "Void",
        "description": "Represents NULL values",
        "fields": null,
        "inputFields": null,
        "interfaces": null,
        "enumValues": null,
        "possibleTypes": null
}
```

In Apollo 1.0, this schema results in the following Swift code being generated, `public typealias Void = String`.

We want to use Apollo codegen to generate an SPM module to house the generated types, but this means that anywhere that imports this module has `Void` changed from referencing `()` to `String`, because this is a public typealias from the generated module:
```swift
/// Represents NULL values
public typealias Void = String
```

Unless I'm missing something more obvious, it seems like the best client-based option is to effectively override this typealias by correcting `Void` back to typealias `Swift.Void` in every module in our codebase, to prefer the conventional Swift version of `Void`.
`public typealias Void = Swift.Void`
Note: I can't change the generated `Void` typealias to `= Swift.Void`, as I then start getting errors such as `Type 'Void' cannot conform to 'OutputTypeConvertible'`.

If there's no better option, I'd encourage reopening this issue until this is addressed in a release, as this doesn't seem like an ideal solution. I feel that this could be better solved by prefixing custom scalars with "Apollo" as they were before the 1.0 update, or IMO the better option - by providing users of the codegen tool the optional ability to choose their own prefix for custom scalars.

**Comment by calvincestari** [comment]
> Unless I'm missing something more obvious, it seems like the best client-based option is to effectively override this typealias by correcting `Void` back to typealias `Swift.Void` in every module in our codebase, to prefer the conventional Swift version of `Void`.

You shouldn't need to do it in every module. Per the header comment in the generated files for custom scalars, they are _not_ overwritten during subsequent codegen passes so the changes you make to them will persist.
```
// @generated
// This file was automatically generated and can be edited to
// implement advanced custom scalar functionality.
//
// Any changes to this file will not be overwritten by future
// code generation execution.
```

> Note: I can't change the generated `Void` typealias to `= Swift.Void`, as I then start getting errors such as `Type 'Void' cannot conform to 'OutputTypeConvertible'`.

This error is being raised because the type for your custom scalar must conform to the [CustomScalarType](https://www.apollographql.com/docs/ios/docc/documentation/apolloapi/customscalartype) protocol. That protocol requires that you code how the GraphQL executor within Apollo iOS transforms your custom scalar to/from a JSON representation, and since it looks like it's simply a string (`"Void"`) it should be a trivial extension. Using `public typealias Void = Swift.Void` is exactly what I recommend you do.

I recommend reading the documentation section on [Custom Scalars](https://www.apollographql.com/docs/ios/custom-scalars) and also the section of the [1.0 Migration Guide](https://www.apollographql.com/docs/ios/migrations/1.0) dedicated to [custom scalars](https://www.apollographql.com/docs/ios/migrations/1.0#custom-scalars).

**Comment by dafurman** [comment]
```swift
public typealias Void = Swift.Void

extension Void: CustomScalarType {} // Non-nominal type 'Void' cannot be extended
```
This is what I run into when trying to follow that path. I can't make `Swift.Void` conform to `CustomScalarType`. I could create my own type (or typealias) for `Void` here, but then I'm back where I started: with the initial problem of the type `Void` confusingly referencing an Apollo type rather than the conventional `Swift.Void` wherever my generated package is imported.

**Comment by calvincestari** [comment]
Ah right, `Void` is a non-nominal type.

**Comment by calvincestari** [comment]
This is a difficult one to work around and I think you're going to need to do the tedious 'fix' for now.
> override this typealias by correcting `Void` back to typealias `Swift.Void` in every module in our codebase

> I feel that this could be better solved by prefixing custom scalars with "Apollo" as they were before the 1.0 update, or IMO the better option - by providing users of the codegen tool the optional ability to choose their own prefix for custom scalars.

I don't think prefixing is our preferred solution for this and it's not so easy once you get into the implementation: to which types should it be applied, I'm sure there are users that would want it for all schema types; if it's for custom scalars only what if you only need it applied to a single custom scalar? The solution we're leaning towards is using a custom GraphQL directive which would preserve the current module/type namespacing but substitute the directive name for the custom scalar defined name.

**Comment by dafurman** [comment]
Yeah I appreciate the quick responses and totally understand that this is probably a tricky edge case!
I do understand there's some [wariness about adding additional configuration options](https://github.com/apollographql/apollo-ios/issues/2598#issuecomment-1289608141) that'd complicate the codegen implementation paths. That said, if there's path for a configuration option that could be used to circumvent a new limitation in Apollo 1.0, I'd personally appreciate it being implemented.

> I don't think prefixing is our preferred solution for this and it's not so easy once you get into the implementation: to which types should it be applied, I'm sure there are users that would want it for all schema types; if it's for custom scalars only what if you only need it applied to a single custom scalar?

My suggestion would be an codegen API design that could look something like this. I could certainly be missing some cases here now, but an enum could be used to provide flexibility for well-defined situations in Apollo updates as-needed:
```Swift
ApolloCodegenConfiguration(
	...
    options: ApolloCodegenConfiguration.OutputOptions(
    	...
        typePrefix: ApolloCodegenConfiguration.TypePrefix( // Optional
            prefix: "MyPrefix",
            applyPrefixTo: .typesNamed(["Void"])
        ),
    )
)

extension ApolloCodegenConfiguration {
    enum TypePrefixApplication {
        case all
        case typesNamed([String])
    }
}
```


> The solution we're leaning towards is using a custom GraphQL directive which would preserve the current module/type namespacing but substitute the directive name for the custom scalar defined name.

If this would resolve the issue of conflicting with non-nominal types, then that'd be great too! Do you think that should be tracked in a newly created issue as this one is currently closed, or that this issue should be reopened?

**Comment by calvincestari** [comment]
> Do you think that should be tracked in a newly created issue as this one is currently closed, or that this issue should be reopened?

I'll reopen this one and add it to the [`1.x` milestone](https://github.com/apollographql/apollo-ios/milestone/72) for now; this issue has lots of useful context in it re. a specific problem definition and possible solutions. It's not likely to be in `1.1` nor `1.2` but possibly `1.3` depending on how workload progresses.

**Comment by AnthonyMDev** [comment]
This is going to be added in #3283. Once that is completed, this can be closed as well.
